internal/babysitter/babysitter.go

```diff
@@ -232,85 +292,87 @@ func (b *Babysitter) StartComponent(req *protos.ComponentToStart) error {
 	}
 
 	// Update the routing info.
-	if req.IsRouted {
-		update := func() {
-			g.routing.Lock()
-			defer g.routing.Unlock()
+	routing := g.routing(req.Component)
+	update := func() {
+		routing.Lock()
+		defer routing.Unlock()
+
+		routing.Val.Replicas = maps.Keys(g.addresses)
+		if req.Routed {
 			assignment := &protos.Assignment{
 				App:          b.dep.App.Name,
 				DeploymentId: b.dep.Id,
 				Component:    req.Component,
 			}
-			assignment = routingAlgo(assignment, g.routing.Val.Replicas)
-			g.routing.Val.Assignments = append(g.routing.Val.Assignments, assignment)
+			routing.Val.Assignment = routingAlgo(assignment, routing.Val.Replicas)
 		}
-		update()
 	}
+	update()
 
-	// Start the colocation group, if it hasn't already started.
 	return b.startColocationGroup(g)
 }
 
-// RegisterReplica implements the protos.EnvelopeHandler interface.
+// RegisterReplica implements the envelope.EnvelopeHandler interface.
 func (b *Babysitter) RegisterReplica(req *protos.ReplicaToRegister) error {
 	g := b.group(req.Group)
-
-	// Update routing.
-	g.routing.Lock()
-	defer g.routing.Unlock()
-	for _, replica := range g.routing.Val.Replicas {
-		if req.Address == replica { // already registered
-			return nil
-		}
-	}
-	g.routing.Val.Replicas = append(g.routing.Val.Replicas, req.Address)
-	for i, assignment := range g.routing.Val.Assignments {
-		g.routing.Val.Assignments[i] = routingAlgo(assignment, g.routing.Val.Replicas)
-	}
-
-	// Add pid.
 	g.mu.Lock()
 	defer g.mu.Unlock()
+
+	// Update addresses and pids.
+	if g.addresses[req.Address] {
```

Comment:
Ditto. Let's try avoid double-locking, and update the addresses/pids here, release the lock, and then update routing.

Reply:
Done.